Feat(Mesh): Migrate CNI configuration docs from Kuma to Kong Mesh (#4944)
Kong/developer.konghq.com: app/mesh/cni.md

---
title: 'Configure the {{site.mesh_product_name}} CNI'
description: 'Install and configure {{site.mesh_product_name}} CNI to enable transparent proxying without requiring privileged init containers.'

content_type: reference
layout: reference

products:
  - mesh
breadcrumbs:
  - /mesh/

tags:
  - kubernetes
  - network

min_version:
  mesh: '2.9'

related_resources:
  - text: Mesh DNS
    url: '/mesh/dns/'
  - text: Transparent proxying
    url: '/mesh/transparent-proxying/'
  - text: Kubernetes annotations and labels
    url: /mesh/annotations/
---

For traffic to flow through {{site.mesh_product_name}}, all inbound and outbound traffic for a service must pass through its sidecar proxy.
The recommended way of accomplishing this is via [transparent proxying](/mesh/transparent-proxying/).

> On Kubernetes, this is handled automatically by the `kuma-init` init container, but it requires elevated privileges. The {{site.mesh_product_name}} CNI is an alternative that removes this requirement from every Pod in the mesh, which can make security compliance easier.

Reviewer: ```suggestion
On Kubernetes, this is handled automatically by the `kuma-init` container, which requires elevated privileges for every Pod in the mesh. The {{site.mesh_product_name}} CNI is an alternative that removes this requirement, which can make security compliance easier.
```

Reviewer: better to explain the problem _first_ then talk about removing rather than mix the two

{:.info}

> > The CNI `DaemonSet` itself requires elevated privileges because it

Reviewer: Can we clarify what level these privileges are at e.g. pod level or mesh level

> writes executables to the host filesystem as `root`.

You can install the CNI using either [kumactl](/mesh/#install-kong-mesh) or [Helm](https://helm.sh/).
The default settings are optimized for OpenShift with Multus. To use {{site.mesh_product_name}} CNI in other environments, follow the installation instructions below.

{:.warning}
> {{site.mesh_product_name}} CNI applies `NetworkAttachmentDefinitions` to applications in any namespace with `kuma.io/sidecar-injection` label.
> To apply `NetworkAttachmentDefinitions` to applications not in a mesh, add the label `kuma.io/sidecar-injection` with the value `disabled` to the namespace.

## CNI architecture

The CNI DaemonSet `{{site.mesh_product_name_path}}-cni` consists of two components:

* A CNI installer
* A CNI binary

The components interact as follows:

{% mermaid %}
flowchart LR
 subgraph s1["conflist"]
        n2["existing CNIs"]
        n3["{{site.mesh_product_name_path}}-cni"]
  end
 subgraph s2["application pod"]
        n4["kuma-sidecar"]
        n5["app-container"]
  end
    A["installer"] -- copy binary and setup conf --> n3
    n3 -- configure iptables --> n4
{% endmermaid %}

> 1. The CNI installer copies the `{{site.mesh_product_name_path}}-cni` binary to the CNI directory on the host. When chained, it also sets up chaining for `{{site.mesh_product_name_path}}-cni` in the CNI conflist file. When chaining is disabled, `{{site.mesh_product_name_path}}-cni` is invoked explicitly as per the Pod manifest.

Reviewer: This sentence is confusing, what does 'When chained' mean ? 

Could this be written as 'When chaining is enabled...' to mirror the following sentence.

### Cilium

Use the following settings to install {{site.mesh_product_name}} CNI in a Cilium-managed cluster:

{% cpinstall cilium %}
cni.enabled=true
cni.chained=true
cni.netDir=/etc/cni/net.d
cni.binDir=/opt/cni/bin
cni.confName=05-cilium.conflist
{% endcpinstall %}

{:.warning}
> * You must set the Cilium configuration value `cni-exclusive` or the corresponding Helm chart value `cni.exclusive` to `false` in order to use Cilium and {{site.mesh_product_name}} together. This is necessary starting with the Cilium v1.14.
> * For Cilium versions older than 1.14, use `{{site.set_flag_values_prefix}}cni.confName=05-cilium.conf` instead of `{{site.set_flag_values_prefix}}cni.confName=05-cilium.conflist`.

{:.info}
> To install the {{site.mesh_product_name}} CNI with Cilium on GKE, follow the [Google - GKE](#google---gke) section.

### Calico

Use the following settings to install {{site.mesh_product_name}} CNI in a Calico-managed cluster:

{% cpinstall calico %}
cni.enabled=true
cni.chained=true
cni.netDir=/etc/cni/net.d
cni.binDir=/opt/cni/bin
cni.confName=10-calico.conflist
{% endcpinstall %}

{:.info}

> > To install the {{site.mesh_product_name}} CNI with Calico on GKE, follow the [Google - GKE](#google---gke) section.

Reviewer: The installation sections do not appear to be ordered, I'd suggest either ordering them alphabetically or moving GKE up so that it immediately follows the two cases where it is referenced.
